### PR TITLE
Hw 18

### PR DESCRIPTION
**Pull request by Alisa411**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,58 @@
...
+This repository contains a utility called **fast_seqs** for working with fastq and fasta sequences.
+
+This Python module provides a set of tools for working with biological sequences, including DNA, RNA, and protein sequences.
+
+## Installation
+
+To install fast_seqs tools you need to clone the git repository using the following command:
+```bash
+git clone git@github.com:Alisa411/fast_seqs.git \
+cd fast_seqs
```
Comment (iam28th) [inline README.md:10]: этого недостаточно чтоб заработало - нужно ещё зависимости через pip поставить

File: HW18_main.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+import os
+import sys
+import io
+import datetime
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from random import choice
+from Bio import SeqIO
+from Bio.SeqUtils import gc_fraction
+from bs4 import BeautifulSoup
```
Comment (iam28th) [inline HW18_main.py:11]:
```suggestion
import datetime
import io
import os
import sys
from abc import ABC, abstractmethod
from dataclasses import dataclass
from random import choice

import requests
from Bio import SeqIO
from Bio.SeqUtils import gc_fraction
from bs4 import BeautifulSoup
```

File: HW18_main.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+        Defines common operations for biological sequences, such as DNA, RNA, or protein sequences.
+
+        Attributes:
+            sequence (str): The sequence of characters representing the biological sequence.
+
+        Methods:
+            __len__(): Returns the length of the biological sequence.
+            __getitem__(index): Gets the character at the specified index or returns a subsequence.
+            __str__(): Returns a string representation of the biological sequence.
```
Comment (iam28th) [inline HW18_main.py:62]: в целом эти дандеры можно наверное было особо не комментировать, они делают плюс-минус то, что от них всегда ожидается (особенно это касается `__str__`)

File: HW18_main.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        return self.sequence[index]
+
+    def complement(self):
+        return type(self)(''.join(self.COMPLEMENT_DICT.get(base, base) for base in self.sequence))
+
+    def gc_content(self, as_percentage=False):
+        gc_count = sum(base in 'GCgc' for base in self.sequence)
+        total_count = len(self.sequence)
+        gc_content = gc_count / total_count if total_count > 0 else 0
+        return gc_content * 100 if as_percentage else gc_content
```
Comment (iam28th) [inline HW18_main.py:133]: в этих классах не хватает type hint-ов

File: HW18_main.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+        'T': 'U',
+        't': 'u'
+    }
+
+    def __init__(self, sequence):
+        super().__init__(sequence)
+
+    def transcribe(self):
+        transcribed_seq = ''.join(self.TRANSCRIBE_DICT.get(base, base) for base in self.sequence)
+        return transcribed_seq
```
Comment (iam28th) [inline HW18_main.py:149]: дело вкуса, но я бы сразу return делал без временной переменной

к слову, прикольное использование второго аргумента в `dict.get` !

File: HW18_main.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+        This function takes  aminoacid sequence and translates in to the RNA.
+        As most of the aminoacids are coded with several different codons,
+        this function will take a random codon of the set for such aminoacids.
+
+        Arguments:
+            seq (str): A sequence of RNA molecule
+
+        Output:
+            returns sequence of aminoacids
+        """
```
Comment (iam28th) [inline HW18_main.py:167]: что-то поплыло

File: HW18_main.py [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+                message = f"🎉 Function `{func.__name__}` successfully finished in `{execution_time_str}`"
+            except Exception as e:
+                message = f"😞 Error in decorator `{func.__name__}`: `{type(e).__name__}: {str(e)}`"
+                send_telegram_message(chat_id, message)
+                raise
+
+        return wrapper
+
+    return decorator
+
```
Comment (iam28th) [inline HW18_main.py:393]: пустая строка не нужна

File: bio_files_processor.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        id (str): The identifier of the record.
+        description (str): The description of the record.
+        seq (str): The sequence data of the record.
+    """
+    id: str
+    description: str
+    seq: str
+
+    def __repr__(self):
+        return f"FastaRecord(id='{self.id}', description='{self.description}', seq='{self.seq}')"
```
Comment (iam28th) [inline bio_files_processor.py:125]: В repr целиком последовательность не очень здорово печатать; если она большая (например, целая хромосома), то такой repr будет не слишком информативным. Посмотрите, как это сделано в Biopython.

File: custom_random_forest.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.base import BaseEstimator
+import numpy as np
+from multiprocessing import Pool
```
Comment (iam28th) [inline custom_random_forest.py:4]:
```suggestion
from multiprocessing import Pool

import numpy as np
from sklearn.base import BaseEstimator
from sklearn.tree import DecisionTreeClassifier
```

File: bio_files_processor.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        raise StopIteration
+
+    def read_record(self):
+        return next(self)
+
+    def read_records(self):
+        records = []
+        for record in self:
+            records.append(record)
+        return records
```
Comment (iam28th) [inline bio_files_processor.py:189]: аннотаций типов не хватает

File: custom_random_forest.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            n_jobs = 1 
+
+        args = [(X, y, np.random.choice(range(n_features), size=self.max_features, replace=False), self.random_state + i)
+                for i in range(self.n_estimators)]
+
+        if n_jobs == 1:
+            results = [self._fit_tree(arg) for arg in args]
+        else:
+            with Pool(n_jobs) as pool:
+                results = pool.map(self._fit_tree, args)
```
Comment (iam28th) [inline custom_random_forest.py:43]: можно лишний раз не if-ать, `Pool` при `n_jobs == 1` должен корректно отрабатывать

File: custom_random_forest.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            probas = np.zeros((X.shape[0], len(self.classes_)))
+            for tree, feat_ids in zip(self.trees, self.feat_ids_by_tree):
+                X_subset = X[:, feat_ids]
+                tree_probas = tree.predict_proba(X_subset)
+                probas += tree_probas
+        else:
+            args = [(tree, feat_ids, X) for tree, feat_ids in zip(self.trees, self.feat_ids_by_tree)]
+            with Pool(n_jobs) as pool:
+                results = pool.map(self._predict_proba_tree, args)
+                probas = np.sum(results, axis=0)
```
Comment (iam28th) [inline custom_random_forest.py:66]: здесь тоже первая if-а визуально прям сильно усложняет код

File: custom_random_forest.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+        probas /= self.n_estimators
+        return probas
+
+    def predict(self, X, n_jobs=None):
+        n_jobs = self.n_jobs if n_jobs is None else n_jobs
+        probas = self.predict_proba(X, n_jobs)
+        predictions = np.argmax(probas, axis=1)
+        return predictions
```
Comment (iam28th) [inline custom_random_forest.py:75]: в этом файле не достаёт аннтотаций типов

File: requirements.txt [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+scipy==1.13.0
+setuptools==68.2.2
+six==1.16.0
+soupsieve==2.5
+threadpoolctl==3.4.0
+tzdata==2024.1
+urllib3==2.2.1
+
+
+
```
Comment (iam28th) [inline requirements.txt:26]: пустые строки ни к чему
и вроде бы здесь много лишнего (кажется, вы не используете pandas или scipy)

File: test_module.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def test_run_genscan_correct_url():
+    """
+    Test if the Genscan URL is accessible.
+    """
+    url = "http://argonaute.mit.edu/cgi-bin/genscanw_py.cgi"
+    response = requests.post(url)
+
+    assert response.status_code == 200
```
Comment (iam28th) [inline test_module.py:79]: здесь либо переименовать тест (мол что он что именно доступность сервера проверяет), либо использовать ту же самую (глобальную) переменную , которая используется в `run_genscan` 

потому из того, что в тесте вы захордкодили правильную ссылку, не обязательно следует, что в другом месте вы захардкодили тоже правильную ссылку)

**Review by iam28th — COMMENTED** [review]
Алиса, добрый день!

### 1. Рефакторинг

Readme и в целом репозиторий весьма опрятные.
Разве что примеры в ноутбуке было бы здорово снадбить пояснениями.

К коду есть ряд претензий: не везде проаннтированы типы, кое-где форматирование поехало, импорты не в правильном порядке; для последнего советую настроить и использовать isort или ruff. Тут -3б

Ещё -1б из-за неточного содержания `requirements.txt`

11/15

### 2. Лес

-1б за усложнение/дублирование кода: лучше запускать pool с одним потоком вместо отдельного if-а.
В остальном всё здорово!

24/25

### 3. Тесты

Тут всё хорошо, только на 2 меньше чем просили в задании, поэтому 8.4/10 баллов.

---

Итого у вас 43.4/50 за это дз, хорошая работа!

Успехов с защитой в ИБ и в дальнейшей карьере!
